### PR TITLE
Add IntField, LongField, FloatField and DoubleField

### LINKED ISSUE #11199: Add IntField, LongField, FloatField and DoubleField classes to index both points and doc values [LUCENE-10162]
Currently we have IntPoint, LongPoint, FloatPoint and DoublePoint on the one hand, and NumericDocValuesField and SortedNumericDocValuesField on the other hand.

When we introduced these classes, this distinction made sense: use the XXXPoint classes if you want your numeric fields to be searchable and the XXXDocValuesField classes if you want your numeric fields to be sortable/aggregatable.

However since then, we introduced logic to take advantage of doc values for filtering (IndexOrDocValuesQuery) and enhanced sorting to take advantage of the Points index to skip non-competitive documents. So even if you only need searching, or if you only need sorting, it's likely a good idea to index both with points **and** doc values.

Could we make this easier on users by having XXXField classes that automatically do it as opposed to requiring users to add both an XXXPoint and an XXXDocValuesField for every numeric field to their index? This could also make consuming these fields easier, e.g. factory methods for range queries could automatically use IndexOrDocValuesQuery.

---
Migrated from [LUCENE-10162](https://issues.apache.org/jira/browse/LUCENE-10162) by Adrien Grand (@jpountz), 1 vote, updated Apr 04 2022
Attachments: [LUCENE-10162-1.patch](https://apache.github.io/lucene-jira-archive/attachments/LUCENE-10162/LUCENE-10162-1.patch)

### PR DESCRIPTION
**Pull request by fcofdez**
This commit adds new IndexableFields that index both points and doc values at once.

Closes #11199

### CODE REVIEW

**Review by jpountz — COMMENTED** [review]
I like it and would be curious to get @rmuir 's thoughts on this PR.

Maybe we could cut over a couple tests that index both points and doc values on the same field to these new `XXXField` classes. I have `TestLongDistanceFeatureQuery`, `TestIndexOrDocValuesQuery` and `TestSortOptimization` in mind, since these tests are about functionality that requires points and doc values to store the same information to work well.

File: lucene/core/src/java/org/apache/lucene/document/DoubleField.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+public final class DoubleField extends Field {
+  /**
+   * Creates a new DoubleField, indexing the provided point and storing it as a DocValue
+   *
+   * @param name field name
+   * @param value the double value
+   * @param sorted configure the field to support multiple DocValues
+   * @throws IllegalArgumentException if the field name or value is null.
+   */
+  public DoubleField(String name, double value, boolean sorted) {
```
Comment (rmuir) [inline lucene/core/src/java/org/apache/lucene/document/DoubleField.java:51]: Should we just make the field multivalued always to simplify it? Or, make the default `true' and provide another ctor with `boolean multivalued`? I think i'd prefer `multivalued` to `sorted` as the name of the parameter.

**Comment by rmuir** [comment]
I like the idea too. Make the .document API simpler for typical use-cases! I added a couple cosmetic comments.

File: lucene/core/src/java/org/apache/lucene/document/DoubleField.java (hunk 1 above)
Comment (gsmiller) [reply lucene/core/src/java/org/apache/lucene/document/DoubleField.java:51]: +1 to the suggestion of always making it multivalued. What if we always set the docvalue type to `SORTED_NUMERIC`, even if the user is just storing single values? Lucene will optimize the storage if all docs are single-valued, and it's already a common pattern to look for this optimization on read (e.g,. https://github.com/apache/lucene/blob/main/lucene/facet/src/java/org/apache/lucene/facet/LongValueFacetCounts.java#L303).

**Comment by jpountz** [comment]
> What if we always set the docvalue type to SORTED_NUMERIC, even if the user is just storing single values? Lucene will optimize the storage if all docs are single-valued, and it's already a common pattern to look for this optimization on read

This sounds good to me. The main issue I see with this is that our default `SortField`s fail if the field is indexed with multi-valued doc values. But maybe we could make it more lenient and sort on the min value for ascending sorts and max values for descending sorts.

**Comment by fcofdez** [comment]
Thanks all for the reviews! 

> This sounds good to me. The main issue I see with this is that our default SortFields fail if the field is indexed with multi-valued doc values. But maybe we could make it more lenient and sort on the min value for ascending sorts and max values for descending sorts.

I noticed this issue while I was running the sorting tests, maybe we can tackle this in a follow up PR? This one is getting big already, but I'm open to suggestions.

**Comment by jpountz** [comment]
Absolutely, adding support for multivalued fields to default sort fields would be a big change too.

**Comment by gsmiller** [comment]
> maybe we can tackle this in a follow up PR?

+1

**Comment by rmuir** [comment]
There is already `SortedSetSortField` and `SortedNumericSortField` that do this. So I don't understand why it would be a big change? Maybe just add some sugar (e.g. `newSortField` or similar)

**Comment by jpountz** [comment]
Good point. I was thinking of making changes to `SortField` but adding factory methods to these Field classes feels like a better approach to me.

**Comment by rmuir** [comment]
OK, well maybe defer it if we have options. I haven't thought about the `SortField` approach, maybe it could be more natural that way.

**Comment by jpountz** [comment]
I just found an issue with this PR which is that multi-valued floats and doubles are encoded with `Double#doubleToLongBits` while they should use `NumericUtil#doubleToSortableLong` to present ordering within a document.

One way to fix it would be to change the encoding depending on whether the field is multi-valued, but this isn't great as we then also need to pass this flag to factor methods for queries (or sort fields in the future). So maybe we should go with @rmuir's earlier suggestion of only supporting multi-valued fields, so that there is a single encoding in doc values?

**Comment by jpountz** [comment]
@fcofdez I hope you don't mind but I checked out your branch to look into the above issue and ended up pushing my changes. These new fields not only support indexing sorted numeric doc values, so that queries do not have to specify whether floats/doubles are indexed via their raw bits or as sortable ints/longs.

**Comment by fcofdez** [comment]
@jpountz thanks for looking into this, unfortunately I didn't have time to go through the last round of review comments. The change makes sense to me.

**Review by fcofdez — COMMENTED** [review]
@jpountz the latest changes LGTM, thanks for moving this forward 👍

**Comment by jpountz** [comment]
Thanks for having a look @fcofdez. I plan on merging this PR soon if there are no objections.

**Review by rmuir — APPROVED** [review]

**Comment by jpountz** [comment]
I cut over the NYC taxis benchmark to these new fields and it yielded a small but consistent speedup. https://home.apache.org/~mikemccand/lucenebench/sparseResults.html#index_throughput

**Pull request merged**